File: Allura/allura/lib/widgets/subscriptions.py
#       Licensed to the Apache Software Foundation (ASF) under one
#       or more contributor license agreements.  See the NOTICE file
#       distributed with this work for additional information
#       regarding copyright ownership.  The ASF licenses this file
#       to you under the Apache License, Version 2.0 (the
#       "License"); you may not use this file except in compliance
#       with the License.  You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#       Unless required by applicable law or agreed to in writing,
#       software distributed under the License is distributed on an
#       "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#       KIND, either express or implied.  See the License for the
#       specific language governing permissions and limitations
#       under the License.

import ew as ew_core
import ew.jinja2_ew as ew
from tg import tmpl_context as c

from allura.lib import validators as V
from allura.lib.widgets import form_fields as ffw
from allura.lib.widgets.forms import CsrfForm
from allura import model as M

from .form_fields import SubmitButton

# Discussion forms


class _SubscriptionTable(ew.TableField):

    class hidden_fields(ew_core.NameList):
        subscription_id = ew.HiddenField(validator=V.Ming(M.Mailbox))
        tool_id = ew.HiddenField()
        project_id = ew.HiddenField()
        topic = ew.HiddenField()
        artifact_index_id = ew.HiddenField()

    class fields(ew_core.NameList):
        project_name = ffw.DisplayOnlyField(
            label='Project', show_label=True, with_hidden_input=False)
        tool = ffw.DisplayOnlyField(
            label='Tool', show_label=True, with_hidden_input=False)
        artifact_title = ew.LinkField(
            label='Item(s)', show_label=True, plaintext_if_no_href=True)
        # unsubscribe = SubmitButton()
        subscribed = ew.Checkbox(suppress_label=True)


class SubscriptionForm(CsrfForm):
    defaults = dict(
        ew.SimpleForm.defaults,
        id='user-subs-form',
        submit_text='Save')

    template = 'jinja:allura:templates/widgets/user_subs_form.html'

    @property
    def fields(self):
        return [
            _SubscriptionTable(name='subscriptions'),
            ew.SingleSelectField(
                name='email_format',
                show_label=False,
                options=[
                    ew.Option(py_value='plain', label='Plain Text'),
                    ew.Option(py_value='both', label='HTML')]),
        ]


class SubscribeForm(ew.SimpleForm):
    template = 'jinja:allura:templates/widgets/subscribe.html'
    defaults = dict(
        ew.SimpleForm.defaults,
        thing='tool',
        style='icon',
        value=None)

    class fields(ew_core.NameList):
        subscribe = SubmitButton()
        unsubscribe = SubmitButton()
        shortname = ew.HiddenField()

    def from_python(self, value, state):
        return value

    def resources(self):
        yield from super().resources()
        if not c.user.is_anonymous():
            yield ew.JSLink('js/subscriptions.js', location='body_js_tail')  # location, to force after react js files

File: Allura/allura/lib/widgets/resources/js/subscriptions.js
/*
       Licensed to the Apache Software Foundation (ASF) under one
       or more contributor license agreements.  See the NOTICE file
       distributed with this work for additional information
       regarding copyright ownership.  The ASF licenses this file
       to you under the Apache License, Version 2.0 (the
       "License"); you may not use this file except in compliance
       with the License.  You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

       Unless required by applicable law or agreed to in writing,
       software distributed under the License is distributed on an
       "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
       KIND, either express or implied.  See the License for the
       specific language governing permissions and limitations
       under the License.
*/

var dom = React.createElement;

/* top-level form state */
var state = {
  'thing': 'tool',
  'subscribed': false,
  'url': '',
  'icon': {}
};

SubscriptionForm = React.createClass({

  getInitialState: function() {
    return {tooltip_timeout: null};
  },

  render: function() {
    var action = this.props.subscribed ? "Unsubscribe from" : "Subscribe to";
    var title = action + ' this ' + this.props.thing;
    var opts = {
      ref: 'link',
      className: this.props.icon.css + (this.props.subscribed ? ' active' : ''),
      href: '#',
      title: title,
      onClick: this.handleClick
    };
    if (this.props.in_progress) {
      opts.style = {cursor: 'wait'};
    }
    return dom('a', opts);
  },

  handleClick: function(e) {
    e.preventDefault();
    var url = this.props.url;
    var csrf = $.cookie('_session_id');
    var data = {_session_id: csrf};
    if (this.props.subscribed) {
      data.unsubscribe = true;
    } else {
      data.subscribe = true;
    }
    set_state({in_progress: true});
    $.post(url, data, function(resp) {
      if (resp.status == 'ok') {
        set_state({
          subscribed: resp.subscribed,
        });
        var text = null;
        var msgStat = null;
        if (resp.subscribed_to_tool) {
          msgStat = 'error';
          text = "You can't subscribe to this ";
          text += this.props.thing;
          text += " because you are already subscribed to the entire ";
          text += resp.subscribed_to_entire_name ? resp.subscribed_to_entire_name : "tool";
          text += ".";
        } else {
          msgStat = 'info';
          var action = resp.subscribed ? 'subscribed to' : 'unsubscribed from';
          text = 'Successfully ' + action + ' this ' + this.props.thing + '.';
        }
        $('#messages').notify(text, {status: msgStat});
      }
    }.bind(this)).always(function() {
      set_state({in_progress: false});
    });
    return false;
  },

  getLinkNode: function() { return ReactDOM.findDOMNode(this.refs.link); },

  componentDidMount: function() {
    var link = this.getLinkNode();
    $(link).tooltipster({
      content: null,
      animation: 'fade',
      delay: 200,
      trigger: 'hover',
      position: 'top',
      iconCloning: false,
      maxWidth: 300
    });
  }

});

function set_state(new_state) {
  /* Set state and re-render entire UI */
  for (var key in new_state) {
    state[key] = new_state[key];
  }
  render(state);
}

function render(state) {
  var props = {};
  for (var key in state) { props[key] = state[key]; }
  ReactDOM.render(
    dom(SubscriptionForm, props),
    document.getElementById('subscription-form')
  );
}

$(function() {
  set_state(document.SUBSCRIPTION_OPTIONS);
});
